flatfs.go

```diff
@@ -404,6 +438,113 @@ func (fs *Datastore) walkTopLevel(path string, reschan chan query.Result) error
 	return nil
 }
 
+// calculateDiskUsage tries to read the diskUsageFile for a cached
+// diskUsage value, otherwise walks the datastore files.
+func (fs *Datastore) calculateDiskUsage() error {
+	// Try to obtain a previously stored value from disk
+	if persDu := fs.readDiskUsageFile(); persDu > 0 {
+		fs.diskUsageCh <- int64(persDu)
+		return nil
+	}
+
+	// Otherwise walk the datastore files
+	var du int64
+	err := filepath.Walk(fs.path,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			du += info.Size()
+			return nil
+		})
+
+	if err != nil {
+		log.Errorf("cannot calculate disk usage: %s", err)
+		return err
+	}
+
+	fs.diskUsageCh <- du
+	return nil
+}
+
+// diskUsageUpdater should run in a goroutine.
+// only call after calculateDiskUsage() has stored
+// a correct value in fs.diskUsage.
+func (fs *Datastore) diskUsageUpdater() {
+	var curDu int64
+
+	for duInc := range fs.diskUsageCh {
+		curDu += duInc // duInc might be < 0
+		atomic.StoreUint64(&fs.diskUsage, uint64(curDu))
+	}
+}
+
+func fileSize(path string) int64 {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return 0
+	}
+	return fi.Size()
+}
+
+// updateDiskUsage reads the size of path and sends it either
+// as an increment or a decrement to the diskUsageCh.
+func (fs *Datastore) updateDiskUsage(path string, add bool) {
+	fsize := fileSize(path)
+	if !add {
+		fsize = -fsize
+	}
+
+	if fsize != 0 {
+		fs.diskUsageCh <- fsize
+	}
+}
+
+func (fs *Datastore) persistDiskUsageFile() {
+	// Store DiskUsage on clean shutdowns.
+	du, err := fs.DiskUsage()
+	if err != nil {
+		// do not store on errors. just ignore them
+		return
+	}
+
+	duB := []byte(fmt.Sprintf("%d", du))
+	ioutil.WriteFile(filepath.Join(fs.path, diskUsageFile), duB, 0644)
```

Comment:
Should really use a tempfile with a rename.

Reply:
can do, can you explain the logic behind it though?

Reply:
btw, my /tmp folder lives in a partition different from my ipfs storage. In the case when you have a system with an ssd and a spinning disk, it is not unusual that /tmp /var etc are mounted on the spinning disk (to reduce the number of writes to the ssd) and the rest of things on the ssd, so imagine the performance hit of current implementation with tmp files and renames..

Reply:
Ah, no. I mean an `ioutil.TempFile` as we do in `doPut`. That is, create a temporary file, write to it, close it, and then atomically rename it into place (so we don't end up with a half-written or corrupted file if we crash while recording the size). Probably not much of an issue in this case but still good practice.

Reply:
yes. i just realized that tmpfiles are created in the same folder. yeah makes sense. Thanks!